### PR TITLE
:sparkles: Add sponsorship process (users + stats)

### PR DESCRIPTION
**Pull request by Jdevvv**
Ca a l'air de bien matché avec la refacto des challenges #36 

### How to test

Run `just dev` & `just test`

### CODE REVIEW

File: functions/data/users.dataset.js [hunk 1]
```diff
@@ -35,4 +32,19 @@ module.exports = [
         team: generateTeam(),
         uid: falso.randUuid(),
     },
-]
+    {
```
Comment (Thisisjuke) [inline functions/data/users.dataset.js:35]: Should create a function named `generateUser` and just export an array `[generateUser(), generateUser()]`

Check if the 2 users are different, if not, check about `seed` from falso

File: functions/modules/stats/index.js [hunk 2]
```diff
@@ -116,120 +117,126 @@ exports.actionUpdate = functions
...
-                }
-            })
-    })
+            .get();
+        if (!statsRef.empty) {
+            await statsRef.docs[0].ref.update(updates);
+        }
+
+        // If user sponsor is updated, increment sponsorCount of sponsoring User
+        if (previousValues.sponsor != newValues.sponsor) {
```
Comment (Thisisjuke) [inline functions/modules/stats/index.js:198]: Need to check if the previous value is null / empty. Logically, we can't change our sponsor : we can't remove badge or decrement the sponsorCount, so it should be impossible to change

File: functions/modules/stats/methods/update-sponsor-count.spec.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    beforeEach(async () => {
+        await deleteCollectionsContent(db, ['users', 'stats']);
+        await db.doc(userPath).set(usersData[1]);
+        await db.doc(statPath).set({
+            ...statsData.emptyStats,
+            uid: usersData[1].uid,
+        });
+    });
+
+    test('error - sponsorship_code invalid', async () => {
```
Comment (Thisisjuke) [inline functions/modules/stats/methods/update-sponsor-count.spec.js:35]: Test name should be more explicit

File: functions/modules/stats/methods/update-sponsor-count.spec.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        const sponsorshipCode = 'sponsorship_code';
+
+        try {
+            await updateSponsorCount(db, sponsorshipCode);
+        } catch (e) {
+            expect(e.message).toBe(`Invalid sponsorship code`);
+        }
+    });
+
+    test('update user sponsorshipCount stats ', async () => {
```
Comment (Thisisjuke) [inline functions/modules/stats/methods/update-sponsor-count.spec.js:45]: Test name should be more explicit

File: functions/modules/stats/methods/update-sponsor-count.spec.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        const sponsorshipCode = usersData[1].sponsorship_code;
+
+        const beforeStat = (await db.doc(statPath).get()).data();
+
+        await updateSponsorCount(db, sponsorshipCode);
+
+        const afterStat = (await db.doc(statPath).get()).data();
+
+        expect(afterStat.sponsorshipCount).toEqual(
```
Comment (Thisisjuke) [inline functions/modules/stats/methods/update-sponsor-count.spec.js:54]: Add another `expect` that should check if the user has the sponsorCode in his document

File: functions/modules/stats/methods/update-sponsor-count.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+const { fieldValue } = require('../../../db-setup');
+
+module.exports.updateSponsorCount = async function (db, sponsorshipCode) {
+    const sponsorSnap = await db
```
Comment (Thisisjuke) [inline functions/modules/stats/methods/update-sponsor-count.js:4]: Reuse the function `checkIfSponsorCodeExists`

**Review by Thisisjuke — CHANGES_REQUESTED** [review]

File: functions/modules/stats/methods/update-sponsor-count.spec.js (hunk 5 above)
Comment (Jdevvv) [reply functions/modules/stats/methods/update-sponsor-count.spec.js:55]: The function `updateSponsorCount` in this test only increment stats `sponsorshipCount` and does not update sponsor code

File: functions/data/users.dataset.js [hunk 7]
```diff
@@ -16,23 +16,26 @@ function generateTarget() {
...
-        last_Name: falso.randLastName(),
-        photo_url: falso.randAvatar(),
-        sponsor: generateSponsorCode(),
-        sponsorship_code: generateSponsorCode(),
-        target: generateTarget(),
-        team: generateTeam(),
-        uid: falso.randUuid(),
-    },
-]
+module.exports.generateUser = (number = 1) => {
```
Comment (Thisisjuke) [inline functions/data/users.dataset.js:19]: generateUser should only generate 1 user

If a loop is needed, the loop will be added outside this function

**Review by Thisisjuke — CHANGES_REQUESTED** [review]

**Pull request merged**